### PR TITLE
fix: Correct cell rotation direction and guard against degenerate cell hashes and image counts

### PR DESCRIPTION
**Pull request by n-gao**
This PR fixes three interconnected bugs that caused incorrect or crashing behavior for non-lower-triangular cell inputs, particularly the FCC primitive 2×2×2 supercell case.

### 1. `to_lower_triangular` rotation direction was inverted

The QR decomposition produces `vecs == L @ Q.T`, meaning the rotation that maps lattice vectors row-by-row into the rows of `L` is right-multiplication by `Q` (i.e., `r_new = r @ Q`). The previous implementation transposed `Q` before returning it, which inverted the rotation direction and caused fractional coordinates to drift whenever the input cell was not already lower-triangular.

### 2. `_cell_hash` could produce out-of-range bin indices

`cell.fold`'s `r % 1` can round up to exactly `1.0` in float arithmetic for inputs slightly below zero. This caused the per-axis bin index to escape `[0, num_cells - 1]`, producing a hash outside `[0, prod(num_cells))` that was then silently dropped by downstream `bincount(length=prod(num_cells))` segment counts, causing atoms to lose all their neighbor candidates.

The fix clamps the bin index with `jnp.clip` and switches the stride computation from `/` to `//` to avoid float precision issues.

### 3. `_get_candidate_images` propagated non-finite ratios into integer image counts

When a cell has a zero or NaN perpendicular length, `cutoff / perp` produces `inf` or `NaN`. Previously the finiteness guard checked `cutoffs` directly rather than the computed `ratio`, so `ceil(inf).astype(int)` wrapped to `INT32_MAX`. This value then propagated through the capacity resize loop and caused a static shape overflow (`OverflowError: 17179869176 out of bounds for int32`) at JAX trace time. The fix computes `ratio` once and applies the `isfinite` guard to it.

## Supporting Changes

- Added `_ensure_lower_triangular` helper in the ASE test utilities to rigidly rotate arbitrary ASE lattices into the lower-triangular convention required by `TriclinicFrame` before passing them to kups, preventing silent projection/truncation of the basis matrix.
- Added the FCC primitive 2×2×2 case to the parametrized neighbor list test suite.
- Added regression tests covering all three bug paths, including direct unit tests for `_cell_hash` boundary clamping, `_get_candidate_images` finite-guard behavior, and fractional-coordinate preservation through `to_lower_triangular`.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#178](https://github.com/cusp-ai-oss/kUPS/pull/178) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#169** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/169?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#168** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/168?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#167** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/167?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#166** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/166?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#165** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/165?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#164** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/164?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#155** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/155?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#154** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/154?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#153** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/153?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#152** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/152?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#151** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/151?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#150** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/150?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#149** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/149?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#148** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/148?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#147** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/147?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#146** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/146?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#145** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/145?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#144** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/144?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#143** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/143?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#142** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/142?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#141** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/141?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#138** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/138?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#137** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/137?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#136** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/136?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#135** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/135?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/135?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#122** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/122?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#121** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/121?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#134** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/134?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#133** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/133?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#132** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/132?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#131** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/131?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#120** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/120?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#119** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/119?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#118** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/118?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/kups/core/cell.py [hunk 1]
```diff
@@ -670,5 +670,13 @@ def to_lower_triangular(vecs: Array) -> tuple[Array, TriclinicMap]:
     R = R * signs[:, None]
     Q = Q * signs[None, :]
     L = R.T
-    Q = Q.T
```
Comment (jonkhler) [inline src/kups/core/cell.py:673]: is this verbosity with inline comments justified? shouldn't that be better documented with a pinning test?

File: src/kups/core/neighborlist.py [hunk 2]
```diff
@@ -257,9 +257,19 @@ def __call__[P: NeighborListPoints](
 
 
 def _cell_hash(coordinate: Array, num_cells: Array):
-    """Calculate the cell hash for a given coordinate and number of cells."""
-    factor = jnp.cumprod(num_cells, axis=-1) / num_cells
-    return (jnp.floor(coordinate * num_cells) * factor).astype(int).sum(axis=-1)
+    """Calculate the cell hash for a given coordinate and number of cells.
+
+    The caller is expected to have folded ``coordinate`` into ``[0, 1)`` per
```
Comment (jonkhler) [inline src/kups/core/neighborlist.py]: i don't think bug fixes should leak into docstrings!

File: src/kups/core/neighborlist.py [hunk 3]
```diff
@@ -464,12 +474,17 @@ def _get_candidate_images(
...
+    ratio = cutoffs[..., None] / perp
     images = jnp.where(
-        cutoffs[..., None] <= perp / 2,
+        ratio <= 0.5,
         1,
-        2 * jnp.ceil(cutoffs[..., None] / perp).astype(int) + 1,
+        2 * jnp.ceil(ratio).astype(int) + 1,
     )
-    images = jnp.where(jnp.isfinite(cutoffs[..., None]), images, 1)
+    # Guard against ``ratio`` being non-finite (degenerate cells with
```
Comment (jonkhler) [inline src/kups/core/neighborlist.py]: same issue as above - I would rather have a documented helper function than long verbose inline comments

**Review by jonkhler — CHANGES_REQUESTED** [review]

**Review by jonkhler — APPROVED** [review]

**Comment by n-gao** [comment]
### Merge activity

* **Jun 2, 12:54 PM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/135).
